### ISSUE TITLE
Failed assertion when ColorScheme.fromSwatch triggered in ThemeData widget with custom primarySwatch

### ISSUE THREAD

**Issue by Laebrye** [flutter/flutter#51857]
Hi,

I've been playing around with app theming to try and get a better understanding of how best to implement consistent themes across a MaterialApp.

Whilst experimenting with the colour and swatch attributes I came across a failed assertion:

```
package:flutter/src/material/color_scheme.dart': Failed assertion: line 36 pos 15: 'primaryVariant
flutter: != null': is not true.

```
The error happened because I did not provide a [700] shade value in the MaterialColor swatch that I provided.

I don't think this is a problem with the ColorScheme.fromSwatch method (although clearly some improved null safety would be useful). I think it might be better to provide the MaterialColor class with a different constructor - one that specifically takes the required integer key values for the map as required parameters.

## Steps to Reproduce

1. Create a new flutter demo counter app
2. Create a custom MaterialColor that lacks the required integer keys (eg. 700 or 200) 
2. Assign the theme with ThemeData that includes the custom MaterialColor assigned to the primarySwatch attribute
3. Run the app

**Expected results:**
When running this, I would have expected to see an error that detailed the missing keys in the MaterialColor swatch or have a constructor for MaterialColor that ensures that required shades are procvided.

**Actual results:**
Failed assertion about the ColorScheme above - which took a little time to figure out the cause when I wasn't getting a compile error.

<details>
  <summary>Logs</summary>

```
[   +3 ms] [DEVICE LOG] 2020-03-03 15:12:22.444337+0000  localhost Runner[75374]: (Flutter) flutter: ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY
╞═══════════════════════════════════════════════════════════
[  +33 ms] [DEVICE LOG] 2020-03-03 15:12:22.445377+0000  localhost Runner[75374]: (Flutter) flutter: The following assertion was thrown building MyApp(dirty):
[        ] [DEVICE LOG] 2020-03-03 15:12:22.445921+0000  localhost Runner[75374]: (Flutter) flutter: 'package:flutter/src/material/color_scheme.dart': Failed assertion: line 36 pos 15:
'primaryVariant
[        ] [DEVICE LOG] 2020-03-03 15:12:22.446690+0000  localhost Runner[75374]: (Flutter) flutter: != null': is not true.
[        ] [DEVICE LOG] 2020-03-03 15:12:22.447300+0000  localhost Runner[75374]: (Flutter) flutter:
[        ] [DEVICE LOG] 2020-03-03 15:12:22.447665+0000  localhost Runner[75374]: (Flutter) flutter: Either the assertion indicates an error in the framework itself, or we should provide
substantially
[        ] [DEVICE LOG] 2020-03-03 15:12:22.448101+0000  localhost Runner[75374]: (Flutter) flutter: more information in this error message to help you determine and fix the underlying
cause.
[        ] [DEVICE LOG] 2020-03-03 15:12:22.448779+0000  localhost Runner[75374]: (Flutter) flutter: In either case, please report this assertion by filing a bug on GitHub:
[        ] [DEVICE LOG] 2020-03-03 15:12:22.449023+0000  localhost Runner[75374]: (Flutter) flutter:   https://github.com/flutter/flutter/issues/new?template=BUG.md
[        ] [DEVICE LOG] 2020-03-03 15:12:22.449727+0000  localhost Runner[75374]: (Flutter) flutter:
[        ] [DEVICE LOG] 2020-03-03 15:12:22.449898+0000  localhost Runner[75374]: (Flutter) flutter: The relevant error-causing widget was:
[        ] [DEVICE LOG] 2020-03-03 15:12:22.450265+0000  localhost Runner[75374]: (Flutter) flutter:   MyApp file:///Users/Dave/AndroidStudioProjects/ratchit/lib/main.dart:5:23
[        ] [DEVICE LOG] 2020-03-03 15:12:22.450847+0000  localhost Runner[75374]: (Flutter) flutter:
[        ] [DEVICE LOG] 2020-03-03 15:12:22.451050+0000  localhost Runner[75374]: (Flutter) flutter: When the exception was thrown, this was the stack:
[        ] [DEVICE LOG] 2020-03-03 15:12:22.451211+0000  localhost Runner[75374]: (Flutter) flutter: #2      new ColorScheme (package:flutter/src/material/color_scheme.dart:36:15)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.451641+0000  localhost Runner[75374]: (Flutter) flutter: #3      new ColorScheme.fromSwatch
(package:flutter/src/material/color_scheme.dart:130:12)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.451931+0000  localhost Runner[75374]: (Flutter) flutter: #4      new ThemeData (package:flutter/src/material/theme_data.dart:287:33)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.452430+0000  localhost Runner[75374]: (Flutter) flutter: #5      MyApp.build (package:ratchit/main.dart:12:14)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.452662+0000  localhost Runner[75374]: (Flutter) flutter: #6      StatelessElement.build (package:flutter/src/widgets/framework.dart:4291:28)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.453291+0000  localhost Runner[75374]: (Flutter) flutter: #7      ComponentElement.performRebuild
(package:flutter/src/widgets/framework.dart:4223:15)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.453654+0000  localhost Runner[75374]: (Flutter) flutter: #8      Element.rebuild (package:flutter/src/widgets/framework.dart:3947:5)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.454153+0000  localhost Runner[75374]: (Flutter) flutter: #9      ComponentElement._firstBuild
(package:flutter/src/widgets/framework.dart:4206:5)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.454517+0000  localhost Runner[75374]: (Flutter) flutter: #10     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4201:5)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.454899+0000  localhost Runner[75374]: (Flutter) flutter: #11     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3194:14)
[  +12 ms] [DEVICE LOG] 2020-03-03 15:12:22.455191+0000  localhost Runner[75374]: (Flutter) flutter: #12     Element.updateChild (package:flutter/src/widgets/framework.dart:2988:12)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.455516+0000  localhost Runner[75374]: (Flutter) flutter: #13     RenderObjectToWidgetElement._rebuild
(package:flutter/src/widgets/binding.dart:1028:16)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.455838+0000  localhost Runner[75374]: (Flutter) flutter: #14     RenderObjectToWidgetElement.mount
(package:flutter/src/widgets/binding.dart:999:5)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.456000+0000  localhost Runner[75374]: (Flutter) flutter: #15     RenderObjectToWidgetAdapter.attachToRenderTree.<anonymous closure>
(package:flutter/src/widgets/binding.dart:942:17)
[  +23 ms] [DEVICE LOG] 2020-03-03 15:12:22.456315+0000  localhost Runner[75374]: (Flutter) flutter: #16     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2412:19)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.456614+0000  localhost Runner[75374]: (Flutter) flutter: #17     RenderObjectToWidgetAdapter.attachToRenderTree
(package:flutter/src/widgets/binding.dart:941:13)
[   +2 ms] [DEVICE LOG] 2020-03-03 15:12:22.456825+0000  localhost Runner[75374]: (Flutter) flutter: #18     WidgetsBinding.attachRootWidget
(package:flutter/src/widgets/binding.dart:819:7)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.457167+0000  localhost Runner[75374]: (Flutter) flutter: #19     WidgetsBinding.scheduleAttachRootWidget.<anonymous closure>
(package:flutter/src/widgets/binding.dart:804:7)
[   +4 ms] [DEVICE LOG] 2020-03-03 15:12:22.457461+0000  localhost Runner[75374]: (Flutter) flutter: #28     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:384:19)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.457600+0000  localhost Runner[75374]: (Flutter) flutter: #29     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:418:5)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.457928+0000  localhost Runner[75374]: (Flutter) flutter: #30     _RawReceivePortImpl._handleMessage
(dart:isolate-patch/isolate_patch.dart:174:12)
[        ] [DEVICE LOG] 2020-03-03 15:12:22.458537+0000  localhost Runner[75374]: (Flutter) flutter: (elided 10 frames from class _AssertionError, package dart:async, and package
dart:async-patch)
[   +6 ms] [DEVICE LOG] 2020-03-03 15:12:22.458691+0000  localhost Runner[75374]: (Flutter) flutter:
[        ] [DEVICE LOG] 2020-03-03 15:12:22.458878+0000  localhost Runner[75374]: (Flutter) flutter:
════════════════════════════════════════════════════════════════════════════════════════════════════
[   +9 ms] [DEVICE LOG] 2020-03-03 15:12:22.798283+0000  localhost Runner[75374]: (FrontBoardServices) [com.apple.BackBoard:Display] [FBSDisplaySource 1-1] silently connecting
<FBSDisplayConfiguration: 0x600003942ca0; Main; mode: "414x896@3x 60Hz p3 SDR">
[        ] [DEVICE LOG] 2020-03-03 15:12:22.798410+0000  localhost Runner[75374]: (FrontBoardServices) [com.apple.BackBoard:Display] [FBSDisplaySource 1-1] initialized
<FBSDisplayConfiguration: 0x600003942ca0; Main; mode: "414x896@3x 60Hz p3 SDR">
[   +5 ms] flutter: ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
[        ] flutter: The following assertion was thrown building MyApp(dirty):
[        ] flutter: 'package:flutter/src/material/color_scheme.dart': Failed assertion: line 36 pos 15: 'primaryVariant
[        ] flutter: != null': is not true.
[  +12 ms] flutter:
[        ] flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
[        ] flutter: more information in this error message to help you determine and fix the underlying cause.
[        ] flutter: In either case, please report this assertion by filing a bug on GitHub:
[        ] flutter:   https://github.com/flutter/flutter/issues/new?template=BUG.md
[        ] flutter:
[        ] flutter: The relevant error-causing widget was:
[        ] flutter:   MyApp file:///Users/Dave/AndroidStudioProjects/ratchit/lib/main.dart:5:23
[        ] flutter:
[        ] flutter: When the exception was thrown, this was the stack:
[        ] flutter: #2      new ColorScheme (package:flutter/src/material/color_scheme.dart:36:15)
[        ] flutter: #3      new ColorScheme.fromSwatch (package:flutter/src/material/color_scheme.dart:130:12)
[   +3 ms] flutter: #4      new ThemeData (package:flutter/src/material/theme_data.dart:287:33)
[        ] flutter: #5      MyApp.build (package:ratchit/main.dart:12:14)
[        ] flutter: #6      StatelessElement.build (package:flutter/src/widgets/framework.dart:4291:28)
[        ] flutter: #7      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4223:15)
[        ] flutter: #8      Element.rebuild (package:flutter/src/widgets/framework.dart:3947:5)
[        ] flutter: #9      ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4206:5)
[        ] flutter: #10     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4201:5)
[        ] flutter: #11     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3194:14)
[        ] flutter: #12     Element.updateChild (package:flutter/src/widgets/framework.dart:2988:12)
[        ] flutter: #13     RenderObjectToWidgetElement._rebuild (package:flutter/src/widgets/binding.dart:1028:16)
[   +1 ms] flutter: #14     RenderObjectToWidgetElement.mount (package:flutter/src/widgets/binding.dart:999:5)
[        ] flutter: #15     RenderObjectToWidgetAdapter.attachToRenderTree.<anonymous closure> (package:flutter/src/widgets/binding.dart:942:17)
[        ] flutter: #16     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2412:19)
[        ] flutter: #17     RenderObjectToWidgetAdapter.attachToRenderTree (package:flutter/src/widgets/binding.dart:941:13)
[        ] flutter: #18     WidgetsBinding.attachRootWidget (package:flutter/src/widgets/binding.dart:819:7)
[        ] flutter: #19     WidgetsBinding.scheduleAttachRootWidget.<anonymous closure> (package:flutter/src/widgets/binding.dart:804:7)
[        ] flutter: #28     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:384:19)
[        ] flutter: #29     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:418:5)
[  +50 ms] flutter: #30     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)
[        ] flutter: (elided 10 frames from class _AssertionError, package dart:async, and package dart:async-patch)
[        ] flutter:
[        ] flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
[  +21 ms] <- compile package:ratchit/main.dart
[+9370 ms] Updating files
[ +773 ms] DevFS: Sync finished
[        ] Syncing files to device iPhone 11 Pro Max... (completed in 10,537ms, longer than expected)
[        ] Synced 8.0MB.
[   +1 ms] Sending to VM service: _flutter.listViews({})
[   +3 ms] Result: {type: FlutterViewList, views: [{type: FlutterView, id: _flutterView/0x7fc96181bc20, isolate: {type: @Isolate, fixedId: true, id: isolates/4499176977989215, name:
main.dart$main-4499176977989215, number: 4499176977989215}}]}
[   +1 ms] <- accept
[        ] Connected to _flutterView/0x7fc96181bc20.
[   +2 ms] 🔥  To hot reload changes while running, press "r". To hot restart (and rebuild state), press "R".
[        ] An Observatory debugger and profiler on iPhone 11 Pro Max is available at: http://127.0.0.1:60782/xtOjT98Q6gw=/
[        ] For a more detailed help message, press "h". To detach, press "d"; to quit, press "q".

```

Analyze output

```
Analyzing ratchit...                                                    
No issues found! (ran in 3.7s)
```

flutter doctor
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.12.13+hotfix.8, on Mac OS X 10.14.6 18G3020, locale en-GB)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
[✓] Android Studio (version 3.6)
[✓] VS Code (version 1.41.1)
[✓] Connected device (1 available)

• No issues found!

```

</details>

**Comment by TahaTesser** [comment]
Hi @Laebrye 

You need to provide complete Material Color Map 
or you can use Material Color maps provided with Flutter 

### Code Sample
```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Material App',
      theme: ThemeData(primarySwatch: Colors.primaries[0]),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      body: Center(
        child: Container(
          child: Text('Hello World'),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () {},
      ),
    );
  }
}
```

Closing, as this isn't an issue with Flutter itself,
if you disagree please write in the comments and I will reopen it 

Thank you

**Comment by Laebrye** [comment]
Hi @TahaTesser,

Thanks for the reply. I'm a bit of a newbie to all this, so I apologise if I'm taking up your time with unnecessary issues.

I raised it because I'm trying to accelerate the app-theming process for my own learning. I got a custom swatch from the tool Material colour palette tool [here](https://material.io/design/color/the-color-system.html#tools-for-picking-colors) 

I wanted to put this generated palette into a custom MaterialColor and apply it to the primarySwatch. I should have read the documentation for MaterialColor first to make sure that I was applying the values correctly - especially in the map - but I was expecting an error or warning _before_ running the app.

If you think this issue should remain closed then I will defer to your judgement. I'll continue working with the theming. I'm going to produce a write-up of everything I come across - if there's any way I can contribute the output to the Flutter project that would be great.

Thanks again.
